### PR TITLE
SCA: Authenticate if necessary when changing subscription payment method

### PR DESCRIPTION
**Pull request by dechov**
Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046

##  Changes proposed

Upon updating the payment method for a subscription, create a SetupIntent, and depending on its status, **redirect to an (SCA) authentication screen** (of the exact sort already in place for the "Pay for order" page) before proceeding with the redirect 👍

Screenshot_2020-12-09 Checkout

Things to note with this implementation:

- The **source is still updated on the subscription when the form is initially submitted**, not after authentication succeeds. This _could_ be avoided by using a filter, but the _notice_ ("Payment method updated.") would still show on the authentication screen page load – see [source](https://github.com/woocommerce/woocommerce-subscriptions/blob/7df33fbbcc089c3d0d7ea0f64765ddfd41dca09e/includes/class-wc-subscriptions-change-payment-gateway.php#L340-L345). I believe it's done that way because that filter is intended for off-site payments (PayPal in particular), where the notice wouldn't actually display until after returning from that flow. _Is it acceptable to proceed with saving the payment method prior to authentication?_
  - If so, then we can consider showing another success message after authentication succeeds. That should be done in [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/832aebe209ceeaf2b8f916dfec6ad5c0916bf3d1/includes/compat/class-wc-stripe-subs-compat.php#L675-L688) (though under what precise conditions is yet TBD).
  - If not, we might need to rework the flow a bit, maybe authenticating from the initial form without a reload (see https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1409)

- There is no error message shown if authentication fails – IIRC it was intended to show after redirect in the "Pay for order" screen case, but that isn't working either, so I think this can be raised outside the scope of the present issue.

- There is also currently no specific error displayed in the front-end if the SetupIntent creation fails (which happened to me somehow when using an old saved card that seems to have been attached to a different Stripe customer) – not sure if that's an issue.

- Instead of always creating a new SetupIntent, we might want to check for an existing SetupIntent and update it, as is done for PaymentIntents [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/832aebe209ceeaf2b8f916dfec6ad5c0916bf3d1/includes/class-wc-gateway-stripe.php#L610-L614)… but I'm not so sure. If a new choice is being made to change the card, I believe that should be a new SetupIntent by definition. If authentication for the same card is being retried, is there a strong reason to avoid creating a new SetupIntent?

- https://github.com/woocommerce/woocommerce-gateway-stripe/commit/0b8d05601923578ad953014820ee2f912be4e7a2 adds hooks for modifying the "Pay for order" screen on an _action_ instead of a _filter_ (so that the hooks can be removed on a subsequent action, in time for [this check](https://github.com/woocommerce/woocommerce-subscriptions/blob/45021ff801f141655c9ab56c0c7636ed89cfda77/includes/class-wc-subscriptions-change-payment-gateway.php#L197) which needs the available gateways). The gateways are still being filtered inside the function – asked about the duplication in https://github.com/woocommerce/woocommerce-gateway-stripe/commit/ec15b62f56bf84b7489ecf75db44954b9b69bd80#r36094893. Is there a reason it was done on a filter that I might have missed? cc @RadoslavGeorgiev 

(I tried to capture the reasons for things here and in commit messages but I'm not sure if I covered them all so of course feel free to ask.)

## Testing instructions

Use instructions from https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046 but **verify that modal authentication appears**, and that if it succeeds then **the renewal also succeeds**. (Just note that the "Edit Subscription" screen needs to be reloaded after the payment method is saved, or else the "Process renewal" action will actually update it back to the earlier source, being the value of the "Stripe Source ID" field in the admin form 🤔.)

(To change the payment method, navigate on the front-end to My account » Subscriptions » any active subscription and click "Change payment", or navigate in wp-admin to WooCommerce » Subscriptions » any active subscription and click "Customer change payment method page →".)

Can also test to make sure the ordinary Pay for Order flow (for regular pending orders) still works as before.

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.
- [x] Did I add a title? A descriptive, yet concise, title.
- [ ] How can this code break?
- [ ] What are we doing to make sure this code doesn't break?

### CODE REVIEW

File: includes/class-wc-gateway-stripe.php
Comment (dechov) [inline includes/class-wc-gateway-stripe.php:939]: Didn't look like this param was used for anything.

File: includes/class-wc-gateway-stripe.php
Comment (DanReyLop) [inline includes/class-wc-gateway-stripe.php:877]: Why did you remove the `wpautop`? It seems like this is being rendered inside a `<li>`, is that the reason why there shouldn't be `<p>`? In that case, the `\n\n` won't render at all, because this is plain HTML, right?

Comment (dechov) [reply includes/class-wc-gateway-stripe.php:877]: The HTML is escaped in the template used for this screen [[source](1405-gh-woocommerce/woocommerce-subscriptions/blob/080fb2e86cb088f02c6a8212d802330baea309c9/templates/checkout/form-change-payment-method.php#L82)], so I thought it might be acceptable to just omit the paragraph break. (Noted in the https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1405/commits/df281efca560ac10b02ea0f6362fe429cdf3e9d1 message.)

I kept them in the string thinking it's not worth impacting the translation – what do you think?

(Taking a step back, another viable approach for the authorization screen might have been to keep Stripe as the one available gateway, and just modify the payment fields output.)

Comment (DanReyLop) [reply includes/class-wc-gateway-stripe.php:877]: > I kept them in the string thinking it's not worth impacting the translation – what do you think?

That's fair, I hadn't read the commit message, I'm fine with that change 👍

**Comment by DanReyLop** [comment]
> Is it acceptable to proceed with saving the payment method prior to authentication?

I think it is acceptable to go with an "optimistic" update.

> There is no error message shown if authentication fails – IIRC it was intended to show after redirect in the "Pay for order" screen case, but that isn't working either, so I think this can be raised outside the scope of the present issue.

I agree, if it's not working for the Pay For Order case, then it shouldn't block this PR.

> If authentication for the same card is being retried, is there a strong reason to avoid creating a new SetupIntent?

Not really a strong reason. It adds some more noise than needed on the Stripe Dashboard, but I don't think it's a big deal.

> which happened to me somehow when using an old saved card that seems to have been attached to a different Stripe customer

This wasn't immediately obvious by looking at the code. Are we creating a SetupIntent even if the card was already saved? That's not needed if #1047 lands.

**Comment by dechov** [comment]
> This wasn't immediately obvious by looking at the code. Are we creating a SetupIntent even if the card was already saved? That's not needed if #1047 lands.

Currently, we're creating a SetupIntent when the payment method is changed, agnostic of whether or not it came from a saved card – which is encapsulated inside the `prepare_source` function. My (possibly incomplete or outdated) understanding was that we won't know exactly when a bank will decide to require authentication, so even if it was authenticated previously, we should still provide a chance to present the challenge again while the customer is on-session.

As for that PR (which I think was https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1407 that you meant to refer to) – I personally think we should either hold that PR while attempting to verify the acceptability of that trade-off (in terms of checkout optimization), or else just leave open the possibility of changing it back at a later point.

**Comment by dechov** [comment]
> This wasn't immediately obvious by looking at the code. Are we creating a SetupIntent even if the card was already saved? That's not needed if #1047 lands.

Currently, we're creating a SetupIntent when the payment method is changed, agnostic of whether or not it came from a saved card – which is encapsulated inside the `prepare_source` function. My (possibly incomplete or outdated) understanding was that we won't know exactly when a bank will decide to require authentication, so even if it was authenticated previously, we should still provide a chance to present the challenge again while the customer is on-session.

As for that PR (which I think was https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1407 that you meant to refer to) – I personally think we should either hold the main change there while attempting to verify the acceptability of that trade-off (in terms of checkout optimization), or else be prepared to revert it at a later point.

**Comment by DanReyLop** [comment]
> which I think was #1407 that you meant to refer to

Yes 🤦 

> I personally think we should either hold the main change there while attempting to verify the acceptability of that trade-off (in terms of checkout optimization), or else be prepared to revert it at a later point.

After seeing your PR, I'm inclined to agree. If a `SetupIntent` is always created before a card is set up for renewals, then there's no need to add the `off_session` flag to everything.

I'll revert the main change and keep the refactors.

Small note: I was looking at how SetupIntents are created (I was worried about Pre-Orders) and turns out that, by default, all SetupIntents setup the card to be used `off_session` ([source](https://stripe.com/docs/api/setup_intents/create#create_setup_intent-usage)). So, in the future, if we consider (or we're told) that being more judicious with that flag will mean less friction for "normal" use cases, we may want to explicitly set it to `on_session` unless dealing with a Subscription/Pre-Order.

> My (possibly incomplete or outdated) understanding was that we won't know exactly when a bank will decide to require authentication, so even if it was authenticated previously, we should still provide a chance to present the challenge again while the customer is on-session.

Yes, that's true, the bank can do whatever they want and nothing we do can change that. That was a good call: if it's "easy" to implement, try to invoke the modal while the user is there.

**Review by DanReyLop — APPROVED** [review]
With the steps you put, it works as expected, and the code looks good. We can ship it as-is.

However, I found a case that's still not covered:
- Use any credit card to sign up for a subscription (`4242424242424242` for example).
- As the same user, go buy a regular (non-subscription) product using the `4000002500003155` card. Check the "save card" checkbox.
- Go to `My Account -> Payment Methods` and set the `*3155` card as default.
- You'll see [this notice](https://user-images.githubusercontent.com/1715800/102019458-9e34fd80-3d6b-11eb-9220-91930f43066c.png), click `Yes`. Note that you don't get an SCA challenge.
- Trigger a renewal. SCA authentication will fail, and the user will get an email telling them to authorize the payment (that part works).

For some reason, clicking `Yes` in that notice goes through [this code path](https://github.com/woocommerce/woocommerce-subscriptions/blob/f10f6d4dd33cab33d34eacdddc5e30040069b0a4/includes/class-wcs-payment-tokens.php#L30), which doesn't trigger the filter you added.

File: includes/class-wc-gateway-stripe.php
Comment (allendav) [inline includes/class-wc-gateway-stripe.php:864]: Why this unusual priority? Please add a comment to the code.

Comment (dechov) [reply includes/class-wc-gateway-stripe.php]: The "Pay for order" form is replaced with "Change payment method" by WooCommerce Subscriptions, on this hook with priority 100 – does the new comment in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1405/commits/6c9833e0def6bd467b04e80f0d07554530dcf3df make that clear enough?

Comment (allendav) [reply includes/class-wc-gateway-stripe.php]: Yes, ty!

**Comment by allendav** [comment]
> Instead of always creating a new SetupIntent, we might want to check for an existing SetupIntent and update it, as is done for PaymentIntents here… but I'm not so sure.

Bear in mind Stripe advises: "Do not maintain long-lived, unconfirmed SetupIntents as they may no longer be valid." https://stripe.com/docs/api/setup_intents

File: assets/js/stripe.js
Comment (allendav) [inline assets/js/stripe.js:755]: Please add a comment along the lines of pi = PaymentIntent, si = SetupIntent

File: includes/class-wc-gateway-stripe.php
Comment (allendav) [inline includes/class-wc-gateway-stripe.php:849]: ❤️ Thank you for making it clear this is an action and not a filter (here and below)

File: includes/class-wc-gateway-stripe.php
Comment (allendav) [inline includes/class-wc-gateway-stripe.php:905]: As an aid in debugging, could we say "SetupIntent" or "PaymentIntent" instead?

File: includes/class-wc-gateway-stripe.php
Comment (allendav) [inline includes/class-wc-gateway-stripe.php:929]: Some older parts of the code say simply intents when they mean (only) PaymentIntents. Could we make it explicit whether this code handles both?

**Review by allendav — CHANGES_REQUESTED** [review]
So, this tests well for me. I get the non payment auth Stripe modal when changing the payment method for a subscription. It does need re-basing and should be re-tested before approval however.

**Comment by v18** [comment]
Great job @dechov getting the update method to work. I having some concerns about the UI:

For context, here's a video of me going through the process:

https://cloudup.com/crl0NWxzE_h

At the 21 second mark, this is right before the SCA modal is shown. The screen looks like this:

Screen Shot 2020-12-17 at 11 50 53 AM

Since we are still waiting for the authentication modal to show up, and the user should not be able to do anything, can we grey out this screen as well? Similar to what shows up right before (~ 19-20 second mark):

Screen Shot 2020-12-17 at 11 51 42 AM

Can we grey out the screen as part of this PR?

I also think it would also be worthwhile to show the success message (the green bar showing at the at top of the page prior to the auth modal) only after the payment is added, specifically on the subscription edit screen that the user gets redirected to. I see that you noted that in your PR description, and if you want to make a new issue for that to fix later that's ok by me 👌 

I think that work will also help with fixing error messaging for when the authentication fails (here's a [video](https://cloudup.com/cW68phDGH6a) showing that for the curious). Let's make an issue for that as well so we don't miss it. Happy for it to go in a later release.

File: includes/class-wc-gateway-stripe.php
Comment (dechov) [reply includes/class-wc-gateway-stripe.php:849]: It was a filter on the available gateways but didn't need to be AFAICT (since those are already being conditionally filtered in this function), so it's been changed to an action in this PR.

**Comment by dechov** [comment]
Thanks all for the testing and reviews – I've realized (a bit late) that **the alternative approach in #1427 (branching off of #1398) is preferable**, for reasons described in that PR body.

Let me know if there's any reason to still favor this solution, but I plan to close this PR if/when #1427 is approved (though one or two of the changes here still seem potentially worthwhile to me, like https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1405/commits/0b8d05601923578ad953014820ee2f912be4e7a2).

@v18 Thanks for emphasizing the problematic success and error messaging in this PR and how the solution in #1398 side-steps the issues. As for "Can we grey out the screen as part of this PR?" – that still seems to me like a good idea for the "Pay for order" screen (the way you'd done [here](https://github.com/Automattic/woocommerce-payments/blob/a71ea77fedf5884085a4de3c8e1e5af84d4de5a2/client/checkout/classic/index.js#L281)), though https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1409 (or some other approach to AJAX intent creation) would remove the page reload entirely.

**Pull request closed without merging**
